### PR TITLE
Improved doc and simplified configuration

### PR DESCRIPTION
**Pull request by moreda**
This set of commits are related to the comments that I wrote in 214189f. After writing the initial zabbix-formula, I lost track of a few commits. At this moment I think it's better to propose these changes (which undo some of the changes already commited by other members of the team) and, after receive some feedback, hopefully commit then :-)

Even at the risk of being repetitive, as I commented in 214189f:

* I think that *mapping all (or most of) the possible values of a key:value configuration file into pillars in a formula has to be carefully considered* because it introduces more lines, more maintenance and fuss, defeating the main purpose of the formula: simplicity for the most frequent cases and base for further private customisation. 

* I don't think that some of the "pillarized" values will be really set with a different value than the default ones in most of cases. Moreover, this strategy leads to key:values in the configuration file that could not be valid for some version of Zabbix (i.e. all the VMware stuff for 2.0.x). In my opinion, the formula should be adapted to complex cases through override of templates and even state files. If not like that, we can end up with a over-complex formula not so really useful and hard to understand. Just as an example, the `pillar.example`section for zabbix server would be something like this:
  ```
  zabbix-server:
    listenip: 0.0.0.0
    listenport: 10051
    dbhost: localhost
    dbname: zabbix
    dbuser: zabbixuser
    dbpass: zabbixpass
    startipmipollers: 0
    startpollersunreachable: 1
    starttrappers: 5
    startpingers: 1
    startdiscoverers: 1
    starthttppollers: 1
    starttimers: 1
    javagatewayport: 10052
    startjavapollers: 0
    startvmwarecollectors: 0
    vmwarefrequency: 60
    vmwarecachesize: 8M
    snmptrapperfile: /tmp/zabbix_traps.tmp
    startsnmptrapper: 0
    housekeepingfrequency: 1
    maxhousekeeperdelete: 500
    senderfrequency: 30
    cachesize: 8M
    cacheupdatefrequency: 60
    startdbsyncers: 4
    historycachesize: 8M
    trendcachesize: 4M
    historytextcachesize: 16M
    valuecachesize: 8M
    nodenoevents: 0
    nodenohistory: 0
    timeout: 3
    trappertimeout: 300
    unreachableperiod: 45
    unavailabledelay: 60
    unreachabledelay: 15
    startproxypollers: 1
    proxyconfigfrequency: 1
    proxydatafrequency: 1
  ```
  instead of the current:
  ```
  zabbix-server:
    listenip: 0.0.0.0
    listenport: 10051
    dbhost: localhost
    dbname: zabbix
    dbuser: zabbixuser
    dbpass: zabbixpass
    extra_conf: |
      # Here we can set extra server configuration lines
  ```
  I honestly think that the second is less scary :-)

* CamelCase notation is not usual in SaltStack/Python/YAML. Yes, it's like that in Zabbix configuration files but e.g. `zabbix-server:HistoryTextCacheSize` is a SaltStack/Python/YAML compound of keys that maps into `HistoryTextCacheSize` Zabbix parameter.

Cheers,

  Roberto

### CODE REVIEW

**Comment by syphernl** [comment]
I commented on 214189f but since it is tucked away I'm quoting myself here to add my 0,02€:
> @moreda I wouldn't completely remove the MySQL service requirement. In most cases it runs on the same server. A better alternative would be introducing a Pillar parameter with the MySQL service name. If the key is not set (or perhaps just "False") it could skip the whole requirement.

> I did have the same issue as @spsoit because our custom MySQL formula used a different name (since we're on MariaDB, not MySQL) but I decided to simply switch to `mysql-server` instead since in the end that's what it is and since this is being relied on by other formulas as well.

> For (bulk) overriding parameters one could use something like [this](https://github.com/saltstack-formulas/dnsmasq-formula/blob/master/dnsmasq/files/dnsmasq.conf#L18) at the end of the configuration file allowing to override the settings. 

> This may not be the cleanest approach since it could potentially add in duplicate entries or keys that are not valid which may cause the daemon to break but it is the most flexible one without having to hardcode all individual entries. When Zabbix 2.6 adds in new configuration entries (quite likely) it doesn't require the formula to be updated.

**Comment by moreda** [comment]
Hi, @syphernl.

> @moreda I wouldn't completely remove the MySQL service requirement. In most cases it runs on the same server.

In this context the main useful effects of keeping the MySQL service requirement is that it's going to raise an error if the state `mysql:service` in not defined and force an state execution order if it does exist. Of course, MySQL is required for a complete Zabbix-MySQL system but it doesn't have to have a strong dependance that **forces** you to: 

1. provision MySQL with Saltstack and 
2. provision MySQL in the minion itself

In all cases that you provision MySQL, you'll need to specify the mysql formula state modules in the top.sls. If you use the right order or (in more sophisticated environments) you use orchestration then you have the clean way to deploy a complete Zabbix-MySQL system.

>  A better alternative would be introducing a Pillar parameter with the MySQL service name. If the key is not set (or perhaps just "False") it could skip the whole requirement.

I would like to minimise pillar or grain values that changes chunks of inside a YAML definition of an state. It's enough with the required Jinja logic that shapes state files to support different OS :-). I'm sure there's a more elegant way to do that, although in this case I strongly push for eliminating the inter-formula requirement.

> I did have the same issue as @spsoit because our custom MySQL formula used a different name (since we're on MariaDB, not MySQL) but I decided to simply switch to mysql-server instead since in the end that's what it is and since this is being relied on by other formulas as well.

This is the type of problems that arise with inter-formula dependencies! :-) ... actually, the funny thing is that at this moment the require is `service: mysql`, so maybe you have to change it again (I didn't do it... I just want to eliminate for all the reasons stated above).

> For (bulk) overriding parameters one could use something like this at the end of the configuration file allowing to override the settings.

When the configuration file is basically `key:value` I understand that it's tempting to model it completely as a dictionary in YAML and then just render it with a macro like the referred one. The problem with generalising this approach is that creates a lot of impedance between what is expected in a template file (something "similar" to a final configuration file with some placeholders for pillar and default values) and what is in for real (something like a non-trivial Jinja macro and a `{%- for name, value in settings.iteritems() %}{{ print_config(name, value) }}{% endfor %}`). This approach also makes more difficult to have a template that "just works" for a basic configuration without basically adding any pillar value and opens the door to too much logic programmed in Jinja. I tried to develop this idea in this [doc](https://github.com/moreda/zabbix-formula/blob/master/TOFS_pattern.md).

I'd vote for using template overriding for specific cases. In Zabbix there's no such variability in the configuration parameters as in dnsmasq (e.g. the `address` dnsmasq` parameter could potentially appear dozens or hundreds of times, so it would make sense to have a list structure in pillar which will be rendered afterwards in a specific position of the configuration file). 

>This may not be the cleanest approach since it could potentially add in duplicate entries or keys that are not valid which may cause the daemon to break but it is the most flexible one without having to hardcode all individual entries. When Zabbix 2.6 adds in new configuration entries (quite likely) it doesn't require the formula to be updated.

I do agree 100% about it not being the cleanest approach :-)
As you can see in my current proposal (in line with the formula that I pushed originally) I have defined just basic minimal pillars for parameters:

1. which values are most likely to be changed in different installations (hostnames, IP, db name, et al)
2. that are supported in Zabbix 1.8, 2.0, 2.2, 2.4 ... and they will likely be in 3.0 (I have insiders info ;-)) OR surrounded by a "if-it's-defined-then-render" wrapper (e.g. HostMetadata).

Sorry for such a long answer, but I'm really interested in a useful generalisation and encapsulation of the formulas. Please, understand that this is my personal approach to the development of formulas. If you don't find anything blatantly wrong I'd like to push the changes and keep maintaining the formula :-)

Cheers!

**Comment by rackergs** [comment]
This might be a bit much to suggest here but why not take an approach like [Augeasproviders](http://augeasproviders.com/) for Puppet here?  Salt does have some [Augeas support](http://docs.saltstack.com/en/latest/ref/modules/all/salt.modules.augeas_cfg.html) already.  This way you can let users directly set items that you dont have to build into the template/config.

**Comment by moreda** [comment]
TL;DR version: I really think that at least in the case of Zabbix, introducing Augeas is overkill :-)

Augeas is really useful when you need to manage complex configurations in a non-destructive way (i.e. when you need to actual "edit" a configuration file). Having the whole configuration file parsed into a tree-like structure that you can modify safely in a programmatic way (look, ma, no sed! :-)) it's really cool... assuming that you have the adequate [lenses](http://augeas.net/docs/lenses.html) for the configuration files involved. I'm going to assume that in order to fully benefit from Augeas we're talking about non-generic lenses.

If you just need to provide a complete configuration file (think of "inmutable configuration files" or created from scratch) then Augeas would just provide another renderer mechanism to add to the list of [Saltstack renderers](http://docs.saltstack.com/en/latest/ref/renderers/) of Saltstack). Actually, a complex one because it requires appropriate and _updated_ lenses :-)

If you use Augeas as a way to provide customisation without hardcoding placeholders in the template, then you're just refining the method based in a Jinja macro + loop that we discussed above. This means bye bye to the simple first approach using minimal templates that resemble real configuration files. Your gain with Augeas would be to ensure a "gramaticaly correct" config file... at the cost of maintaining non-generic lenses adapted to each version of the software.

For instance, you'll have something like this (not really tested, just off the top of my head):

```
zabbix-agent-conf:
  augeas,change:
    - context: /files/etc/zabbix/zabbix_agent.conf
    - changes:
{%- for key, value in params.iteritems() %}
      - set {{ key }} {{ value }}
{%- endfor %}
``` 

Notice that, as you are actually editing a preexisting configuration file, you are just modifying a subset of configuration keys (this could be something not desirable and it could force you to "clean" the file with a `file` state... maybe just to create an empty file).

In my times with Puppet I didn't have any remote execution mechanism to allow changes in an already provisioned host (other than executing again a modified manifest, which is actually "re-provisioning"). My interest in Augeas by then was related to the need of "mutate" configuration files in a better way (no sed) using manifest execution. In all my cases, in the end it was always better to consider them inmutable and not edit them.

Cheers!

**Comment by rackergs** [comment]
lacking deeper knowledge of Salt I don't know that I can properly portray how this work, so we can have this conversation in a different medium.

I can say that augeasproviders in puppet is way nicer than just using augeas directly.

Your example is more like puppet's

```puppet
augeas { 'set zabbix agent':
  context => '/files/etc/zabbix/zabbix_agent.conf',
  changes => ['set blah x', 'set meh y']
}
```

and augeas providers is more:

```puppet
zabbix_agent { 'blah': value => 'x' }
zabbix_agent { 'meh': value => 'y' }
```

**Comment by whiteinge** [comment]
I second not requiring dependencies like Augeas in Formulas unless necessary. Or perhaps making them value-adds that aren't required for the base functionality.

**Comment by moreda** [comment]
Hi,
As I'd like to continue being responsible of the development of the formula, I'm going to merge my pull request. I think that support for Augeas or extra dependencies should be added in a possible fork o specific local branch.
